File: Assets/Scripts/UI/Logic/DpsUI.cs
using LiteFramework.Game.UI;
using LiteFramework.Helper;
using LiteMore.Player;
using UnityEngine;
using UnityEngine.UI;

namespace LiteMore.UI.Logic
{
    public class DpsUI : BaseUI
    {
        private GameObject ItemObj_;
        private Transform ContentRoot_;
        private Text DpsText_;

        private const float Interval_ = 0.5f;
        private float Time_;

        public DpsUI()
            : base()
        {
            DepthMode = UIDepthMode.Normal;
            DepthIndex = 100;
        }

        protected override void OnOpen(params object[] Params)
        {
            ItemObj_ = FindChild("Item").gameObject;
            ItemObj_.SetActive(false);

            ContentRoot_ = FindChild("List/Viewport/Content");

            DpsText_ = FindComponent<Text>("Dps");

            AddEventToChild("BtnClear", () =>
            {
                PlayerManager.Dps.Clear();
                UIHelper.RemoveAllChildren(ContentRoot_);
            });
        }

        protected override void OnTick(float DeltaTime)
        {
            Time_ += DeltaTime;

            if (Time_ >= Interval_)
            {
                Time_ -= Interval_;
                Refresh();
            }
        }

        private void Refresh()
        {
            var Chunks = PlayerManager.Dps.GetChunks();

            for (var Index = 0; Index < ContentRoot_.childCount; ++Index)
            {
                ContentRoot_.GetChild(Index).gameObject.SetActive(false);
            }

            for (var Index = 0; Index < Chunks.Count; ++Index)
            {
                var Child = Index >= ContentRoot_.childCount ? CreateItem() : ContentRoot_.GetChild(Index);

                Child.gameObject.SetActive(true);
                UIHelper.FindComponent<Text>(Child, "Name").text = Chunks[Index].SourceName;
                UIHelper.FindComponent<Text>(Child, "Text").text = $"{Chunks[Index].Value:0.0}({Chunks[Index].Percent*100:0.00}%)";
                UIHelper.FindComponent<Slider>(Child, "Value").value = Chunks[Index].Percent;
            }

            DpsText_.text = $"Dps:{PlayerManager.Dps.Dps}";
        }

        private Transform CreateItem()
        {
            var Obj = Object.Instantiate(ItemObj_);
            Obj.transform.SetParent(ContentRoot_, false);
            UIHelper.FindComponent<Image>(Obj.transform, "Value/Fill Area/Fill").color = new Color(Random.value, 0.8f, Random.value);
            return Obj.transform;
        }
    }
}

File: Assets/StandaloneAssets/scripts/logic/ui/dpsui.lua
local DpsUI = BaseClass('DpsUI', BaseUI)

function DpsUI:Ctor()
	self.DepthMode = UIDepthMode.Normal
	self.Depth = 100

	--[[RegisterCSEvent('NpcAttrChangedEvent', self, function(_, evt)
	end)]]

	self.Interval_ = 0.5
    self.Time_ = 0
end

function DpsUI:Dtor()
	--UnRegisterAllCSEvent(self)
end

function DpsUI:OnOpen()
	ItemObj_ = self:FindChild("Item").gameObject;
    ItemObj_:SetActive(false)

    ContentRoot_ = self:FindChild("List/Viewport/Content")
    DpsText_ = self:GetComponent("Dps", UIText);

    self:AddEventToChild("BtnClear", function()
        CS.LiteMore.Player.PlayerManager.Dps:Clear()
        UIHelper.RemoveAllChildren(ContentRoot_)
    end)

	self:Refresh()
end

function DpsUI:OnClose()
end

function DpsUI:OnTick(dt)
	self.Time_ = self.Time_ + dt

    if self.Time_ >= self.Interval_ then
        self.Time_ = self.Time_ - self.Interval_
        self:Refresh()
    end
end

function DpsUI:Refresh()
	local Chunks = CS.LiteMore.Player.PlayerManager.Dps:GetChunks()

	for i = 0, ContentRoot_.childCount - 1 do
		ContentRoot_:GetChild(i).gameObject:SetActive(false)
	end

    for i = 0, Chunks.Count - 1 do
        local Child
        if i >= ContentRoot_.childCount then
        	Child = self:CreateItem()
        else
        	Child = ContentRoot_:GetChild(i)
        end

        Child.gameObject:SetActive(true);
        UIHelper.GetComponent(Child, "Name", UIText).text = Chunks[i].SourceName
        UIHelper.GetComponent(Child, "Text", UIText).text = string.format("%0.2f", Chunks[i].Value) .. "(" .. string.format("%0.2f", Chunks[i].Percent*100) .. "%)"
        UIHelper.GetComponent(Child, "Value", UISlider).value = Chunks[i].Percent
    end

    DpsText_.text = "Dps:" .. string.format("%0.2f", CS.LiteMore.Player.PlayerManager.Dps.Dps)
end

function DpsUI:CreateItem()
    local Obj = CS.UnityEngine.Object.Instantiate(ItemObj_)
    Obj.transform:SetParent(ContentRoot_, false)
    UIHelper.GetComponent(Obj.transform, "Value/Fill Area/Fill", UIImage).color = CS.UnityEngine.Color(CS.UnityEngine.Random.value, 0.8, CS.UnityEngine.Random.value)
    return Obj.transform
end

return DpsUI